src/sock/socket_session.cpp

```diff
@@ -168,6 +204,34 @@ void socket_session<T>::on_read(error_code ec, std::size_t)
         });
 }
 
+/*
+* Checks whether threshold counters have exceed their max threshold values and if exceeds hand over the rest of work to socket_monitor
+* Initially timestamp is set to 0. So in intial checkup timestamp variable is assigned current timestamp and pass it. From second 
+* instance when this function gets hit it will check whether the timestamp difference between the current and last read is higher
+* than the interval and counter has exceeds the threshold limit and pass it to socket_monitor.
+*/
+template <class T>
+void socket_session<T>::threshold_validator(util::SESSION_THRESHOLDS threshold_key, session_thresholds &threshold)
+{
+    if (threshold.timestamp == 0)
+    {
+        threshold.timestamp = util::get_epoch_milliseconds();
+    }
+    else
+    {
+        auto elapsed_time = util::get_epoch_milliseconds() - threshold.timestamp;
+        if (elapsed_time <= threshold.interval_in_ms && threshold.counter_value > threshold.threshold_limit)
+        {
+            threshold_monitor(threshold_key, threshold.threshold_limit, this);
+        }
+        else if (elapsed_time > threshold.interval_in_ms)
+        {
+            threshold.timestamp = util::get_epoch_milliseconds();
+            threshold.counter_value = 0;
```

Comment:
We should not reset the counter here. This would ignore the counter increment that just happened.